contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,228 @@
+    case 'ADD_TODO':
+      return state.concat([{ text: action.text, completed: false }])
+    case 'TOGGLE_TODO':
+      return state.map((todo, index) =>
+        action.index === index
+          ? { text: todo.text, completed: !todo.completed }
+          : todo
+      )
+    default:
+      return state
+  }
+}
+```
+
+These two reducers are called by a main reducer, and with that the complete app state is managed.
+
+```js
+function todoApp(state = {}, action) {
+  return {
+    todos: todos(state.todos, action),
+    visibilityFilter: visibilityFilter(state.visibilityFilter, action)
+  }
+}
+```
+
+<!-- put a tooltip on pure -->
+The reducers are **pure** functions. They take the previous state and an action, and return the next state, as new state objects, instead of mutating the previous state.
+
+A careful eye would notice that the above code block examples on reducers are not pure functions. If you're interested, you may see what the reducer should actually be written. It includes some Redux API, and Javascript methods such as `.map` and the spread operator to make the functions pure.
+
+## Why use Redux?
+
+<!-- Some points from this section were adapted from these article  box-->
+
+<box type='definition' header="Some points given in this section were adapted from these articles">
+
+* [Why use Redux? Reasons with clear examples by Neo Ighodaro](https://blog.logrocket.com/why-use-redux-reasons-with-clear-examples-d21bffd5835/)
+* [Redux · An Introduction by Alex Bachuk](https://www.smashingmagazine.com/2016/06/an-introduction-to-redux/)
+</box>
+
+Now that we know what Redux is, let us take a look at how the structure of Redux and its concepts make possible some benefits.
+
+<!-- confusing point, make simpler and tie to my article -->
+1. **State is made predictable**<br/>
+In Redux, the single immutable store is the one source of truth for application state. Neither the view of the application, nor network callbacks directly write to it. Changes to the state are via actions, which Redux centralizes and keep in strict order. Also, reducers are pure functions, always producing the same result for the same state and action. Because of all of this, there is a structured and predictable cycle of which applicable state is updated. State is made predictable. Having such a a predictable state, makes it possible to implement functionality that is otherwise difficult, such as Undo and Redo.
+
+2. **Easier to debug and inspect**<br/>
+Since actions and state are plain objects, it is possible to serialize them and log them. In Redux, especially with the Redux DevTools, you can view a log of actions that took place, making it easy to understand errors in code. You can also print the store within the code itself (for example using `console.log(store.getState())`{.js}). The Redux DevTools, used in the browser, also provides features such as a tree visualization of the application state, as well a time travel ability to replay state changes.
```

Comment:
instead of (for example using `console.log...`), can place the example as a tooltip?